src/hotspot/share/opto/predicates.hpp

```diff
@@ -652,52 +660,86 @@ class InitializedAssertionPredicateCreator : public StackObj {
 
 // This class iterates through all predicates of a Regular Predicate Block and applies the given visitor to each.
 class RegularPredicateBlockIterator : public StackObj {
-  Node* const _start_node;
+  Node* _current_node;
   const Deoptimization::DeoptReason _deopt_reason;
 
  public:
   RegularPredicateBlockIterator(Node* start_node, Deoptimization::DeoptReason deopt_reason)
-      : _start_node(start_node),
+      : _current_node(start_node),
         _deopt_reason(deopt_reason) {}
   NONCOPYABLE(RegularPredicateBlockIterator);
 
   // Skip all predicates by just following the inputs. We do not call any user provided visitor.
-  Node* skip_all() const {
+  Node* skip_all() {
     PredicateVisitor do_nothing; // No real visits, just do nothing.
     return for_each(do_nothing);
   }
 
   // Walk over all predicates of this block (if any) and apply the given 'predicate_visitor' to each predicate.
   // Returns the entry to the earliest predicate.
-  Node* for_each(PredicateVisitor& predicate_visitor) const {
-    Node* current = _start_node;
+  Node* for_each(PredicateVisitor& predicate_visitor) {
     while (predicate_visitor.should_continue()) {
-      if (TemplateAssertionPredicate::is_predicate(current)) {
-        TemplateAssertionPredicate template_assertion_predicate(current->as_IfTrue());
-        predicate_visitor.visit(template_assertion_predicate);
-        current = template_assertion_predicate.entry();
-      } else if (RuntimePredicate::is_predicate(current, _deopt_reason)) {
-        RuntimePredicate runtime_predicate(current->as_IfProj());
-        predicate_visitor.visit(runtime_predicate);
-        current = runtime_predicate.entry();
-      } else if (InitializedAssertionPredicate::is_predicate(current)) {
-        InitializedAssertionPredicate initialized_assertion_predicate(current->as_IfTrue());
-        predicate_visitor.visit(initialized_assertion_predicate);
-        current = initialized_assertion_predicate.entry();
-      } else {
-        // Either a Parse Predicate or not a Regular Predicate. In both cases, the node does not belong to this block.
-        break;
+      if (process_template_assertion_predicate(predicate_visitor)) {
+        continue;
+      }
+
+      if (process_runtime_predicate(predicate_visitor)) {
+        continue;
+      }
+
+      if (process_initialized_assertion_predicate(predicate_visitor)) {
+        continue;
       }
```

Comment:
Refactoring: Split into methods and each method check if the opaque node is marked useless before visiting.